File: 50.pow-x-n.js
/*
 * @lc app=leetcode id=50 lang=javascript
 *
 * [50] Pow(x, n)
 */

// @lc code=start
/**
 * @param {number} x
 * @param {number} n
 * @return {number}
 */
var myPow = function (x, n) {
  let res = 1;
  for (let i = Math.abs(n); i !== 0; i = Math.floor(i / 2)) {
    if (i % 2 !== 0) res *= x;
    x *= x;
  }
  return n < 0 ? 1 / res : res;
};
// @lc code=end

File: 50.pow-x-n.py
#
# @lc app=leetcode id=50 lang=python3
#
# [50] Pow(x, n)
#

# Time:  O(logn)
# Time:  O(logn)

# @lc code=start
class Solution:
    def myPow(self, x: float, n: int) -> float:
        if n == -1:
            return 1 / x
        if n == 0:
            return 1
        half = self.myPow(x, n // 2)
        if n % 2 == 0:
            return half * half
        return half * half * x


# @lc code=end
